pkg/scheduler/queue/queue.go

```diff
@@ -245,41 +248,37 @@ func (q *RequestQueue) tryDispatchRequest(broker *queueBroker, call *nextRequest
 	}
 
 	call.lastUserIndex.last = idx
-	if queue == nil {
+	if req == nil {
 		// Nothing available for this querier, try again next time.
 		return false
 	}
 
-	// Pick next request from the queue. The queue is guaranteed not to be empty because we remove empty broker.
-	queueElement := queue.Front()
-
-	requestSent := call.send(nextRequestForQuerier{
-		req:           queueElement.Value,
+	reqForQuerier := nextRequestForQuerier{
+		req:           req,
 		lastUserIndex: call.lastUserIndex,
 		err:           nil,
-	})
+	}
+	requestSent := call.send(reqForQuerier)
 
 	if requestSent {
-		// If GetNextRequestForQuerier received the request, remove it from the queue.
-		// (GetNextRequestForQuerier might have already returned if its context was cancelled.)
-		queue.Remove(queueElement)
-
-		if queue.Len() == 0 {
-			broker.deleteQueue(tenantID)
-		}
-
```

Comment:
Primary change is here.
Access to the queue structure itself is now encapsulated by the `queueBroker`, in order to pave the way for the added complexity of managing the upcoming multi-dimensional queueing structures.

Instead of the consumer being exposed explicitly to a `*list.List` structure and attempting to send the request  to the querier before choosing whether or not to dequeue, the client receives a dequeued request and then can elect to re-push it to the front of the queue depending on the success or failure of sending it to the querier.

This also hides the logic on when to delete empty queues from the consumer, and the consumer does not need to use or know about any behavior specific to the `list.List`.